build-data-capturing-gradle-samples/capture-thermal-throttling/gradle-thermal-throttling.gradle

```diff
@@ -0,0 +1,139 @@
+import org.gradle.api.services.BuildService
+import org.gradle.api.services.BuildServiceParameters
+import org.gradle.internal.os.OperatingSystem
+
+import java.nio.charset.Charset
+import java.util.concurrent.ConcurrentLinkedQueue
+import java.util.concurrent.Executors
+import java.util.concurrent.TimeUnit
+import java.util.function.Function
+
+/**
+ * This Gradle script captures the thermal throttling level and adds it as a tag.
+ * Some parameters can be tweaked according to your build:
+ * - SAMPLING_INTERVAL_IN_SECONDS, frequency at which the capture command is run
+ * - THROTLLING_LEVEL, list of throttling levels and value ranges (to be compared with captured throttling average value)
+ *
+ * WARNINGS:
+ * - This is supported on MacOS only.
+ *
+ * This should be applied to the root project:
+ * <code> apply from: file('gradle-thermal-throttling.gradle') </code>
+ */
+
+def buildScanApi = project.extensions.findByName('buildScan')
+if (!buildScanApi) {
+  return
+}
+
+// Register Thermal throttling service
+def throttlingServiceProvider = gradle.sharedServices.registerIfAbsent("thermalThrottling", ThermalThrottlingService.class) {}
+
+// Start Thermal throttling service
+throttlingServiceProvider.get()
+
+// Process results on build completion
+buildScan {
+  buildFinished { def result ->
+    throttlingServiceProvider.get().processResults(buildScanApi)
+  }
+}
+
+// Thermal Throttling service implementation
+abstract class ThermalThrottlingService implements BuildService<BuildServiceParameters.None>, AutoCloseable {
+
+  /**
+   * Sampling interval can be adjusted according to total build time.
+   */
+  def static final SAMPLING_INTERVAL_IN_SECONDS = 5
+
+  /**
+   * Throttling levels by throttling average value.
+   */
+  def static final THROTTLING_LEVEL =
+          [
+                  [level: "THROTTLING_HIGH", range: 0..40],
+                  [level: "THROTTLING_MEDIUM", range: 40..80],
+                  [level: "THROTTLING_LOW", range: 80..100]
+          ]
+
+  def static final COMMAND_ARGS = ["pmset", "-g", "therm"]
+  def static final COMMAND_OUTPUT_PARSING_PATTERN = /CPU_Speed_Limit\s+=\s+/
+
+  def static final scheduler = Executors.newScheduledThreadPool(1)
+  def static final samples = new ConcurrentLinkedQueue<Integer>()
+
+  static {
+    if (OperatingSystem.current().macOsX) {
+      scheduler.scheduleAtFixedRate(new ProcessRunner(COMMAND_ARGS, ThermalThrottlingService::processCommandOutput), 0, SAMPLING_INTERVAL_IN_SECONDS, TimeUnit.SECONDS)
```

Comment:
@ghale it feels to me that having the scheduler start inside a `static` initializer block might cause some situations where this is started but never closed. Is that possible or is there a better way to do this like inside the creation of this class?

Reply:
I suppose someone could modify the script and create such a situation.  I think it was originally written with the scheduler started in the constructor.  @jprinet was there a particular reason this was moved inside of a static block?  If not, I'd say just move it back to the constructor.

Reply:
I wanted to make sure we would start to capture data as early as possible.
Just to be on the same page here, what's the concern? Encountering an error during object creation and having the scheduler already started?

Reply:
I moved the block to the constuctor which has the side benefit of allowing methods to be non static